src/bloqade/pyqrack/squin/runtime.py

```diff
@@ -0,0 +1,323 @@
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        m = [not self.to_state, 0, 0, self.to_state]
+        target = qubits[-1]
+        ctrls = [qbit.addr for qbit in qubits[:-1]]
+        target.sim_reg.mcmtrx(ctrls, m, target.addr)
+
+
+@dataclass(frozen=True)
+class IdentityRuntime(OperatorRuntimeABC):
+    # TODO: do we even need sites? The apply never does anything
+    sites: int
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        pass
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        pass
+
+
+@dataclass(frozen=True)
+class MultRuntime(OperatorRuntimeABC):
+    lhs: OperatorRuntimeABC
+    rhs: OperatorRuntimeABC
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        if adjoint:
+            # NOTE: inverted order
+            self.lhs.apply(*qubits, adjoint=adjoint)
+            self.rhs.apply(*qubits, adjoint=adjoint)
+        else:
+            self.rhs.apply(*qubits)
+            self.lhs.apply(*qubits)
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        if adjoint:
+            self.lhs.control_apply(*qubits, adjoint=adjoint)
+            self.rhs.control_apply(*qubits, adjoint=adjoint)
+        else:
+            self.rhs.control_apply(*qubits, adjoint=adjoint)
+            self.lhs.control_apply(*qubits, adjoint=adjoint)
+
+
+@dataclass(frozen=True)
+class KronRuntime(NonBroadcastableOperatorRuntimeABC):
+    lhs: OperatorRuntimeABC
+    rhs: OperatorRuntimeABC
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        self.lhs.apply(qubits[0], adjoint=adjoint)
+        self.rhs.apply(qubits[1], adjoint=adjoint)
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        # FIXME: this feels a bit weird and it's not very clear semantically
+        # for now I'm settling for: apply to qubits if ctrls, using the same ctrls
+        # for both targets
+        assert len(qubits) > 2
+        target1 = qubits[-2]
+        target2 = qubits[-1]
+        ctrls = qubits[:-2]
+        self.lhs.control_apply(*ctrls, target1, adjoint=adjoint)
+        self.rhs.control_apply(*ctrls, target2, adjoint=adjoint)
+
+
+@dataclass(frozen=True)
+class ScaleRuntime(OperatorRuntimeABC):
+    op: OperatorRuntimeABC
+    factor: complex
+
+    def mat(self, adjoint: bool):
+        if adjoint:
+            return [np.conj(self.factor), 0, 0, self.factor]
+        else:
+            return [self.factor, 0, 0, self.factor]
+
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        self.op.apply(*qubits, adjoint=adjoint)
+
+        target = qubits[-1]
+
+        # NOTE: just factor * eye(2)
+        m = self.mat(adjoint)
+
+        # TODO: output seems to always be normalized -- no-op?
+        target.sim_reg.mtrx(m, target.addr)
```

Comment:
if this is true we might want to error instead of allowing things to run. If we want to do something like `rot(scale * op)` we can rewrite this to a standard rotation statement. 

cc: @Roger-luo